### PR TITLE
[SPARK-40149][SQL] Propagate metadata columns through Project

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This PR fixes a regression caused by https://github.com/apache/spark/pull/32017 .

In https://github.com/apache/spark/pull/32017 , we tried to be more conservative and decided to not propagate metadata columns in certain operators, including `Project`. However, the decision was made only considering SQL API, not DataFrame API. In fact, it's very common to chain `Project` operators in DataFrame, e.g. `df.withColumn(...).withColumn(...)...`, and it's very inconvenient if metadata columns are not propagated through `Project`.

This PR makes 2 changes:
1. Project should propagate metadata columns
2. SubqueryAlias should only propagate metadata columns if the child is a leaf node or also a SubqueryAlias

The second change is needed to still forbid weird queries like `SELECT m from (SELECT a from t)`, which is the main motivation of https://github.com/apache/spark/pull/32017 .

After propagating metadata columns, a problem from https://github.com/apache/spark/pull/31666 is exposed: the natural join metadata columns may confuse the analyzer and lead to wrong analyzed plan. For example, `SELECT t1.value FROM t1 LEFT JOIN t2 USING (key) ORDER BY key`, how shall we resolve `ORDER BY key`? It should be resolved to `t1.key` via the rule `ResolveMissingReferences`, which is in the output of the left join. However, if `Project` can propagate metadata columns, `ORDER BY key` will be resolved to `t2.key`.

To solve this problem, this PR only allows qualified access for metadata columns of natural join. This has no breaking change, as people can only do qualified access for natural join metadata columns before, in the `Project` right after `Join`. This actually enables more use cases, as people can now access natural join metadata columns in ORDER BY. I've added a test for it.

### Why are the changes needed?

fix a regression

### Does this PR introduce _any_ user-facing change?

For SQL API, there is no change, as a `SubqueryAlias` always comes with a `Project` or `Aggregate`, so we still don't propagate metadata columns through a SELECT group.

For DataFrame API, the behavior becomes more lenient. The only breaking case is an operator that can propagate metadata columns then follows a `SubqueryAlias`, e.g. `df.filter(...).as("t").select("t.metadata_col")`. But this is a weird use case and I don't think we should support it at the first place.

### How was this patch tested?

new tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @karenfeng @viirya @huaxingao

**Review by viirya — COMMENTED** [review]
Looks good. But there are a few SQL test failures. Are they related?

**Comment by cloud-fan** [comment]
@viirya I've updated the PR description for the new issue discovered.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala [hunk 1]
```diff
@@ -191,20 +191,20 @@ package object util extends Logging {
     /**
      * If set, this metadata column is a candidate during qualified star expansions.
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala:192]: This comment is out-of-dated.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -3523,8 +3525,8 @@ class Analyzer(override val catalogManager: CatalogManager)
     val project = Project(projectList, Join(left, right, joinType, newCondition, hint))
     project.setTagValue(
       Project.hiddenOutputTag,
-      hiddenList.map(_.markAsSupportsQualifiedStar()) ++
-        project.child.metadataOutput.filter(_.supportsQualifiedStar))
+      hiddenList.map(_.markAsQualifiedAccess()) ++
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3528]: The new semantics read a bit weird. Previously it is understandable that `markAsSupportsQualifiedStar` means `hiddenList` can be accessed by qualified star.

But how to interpret `markAsQualifiedAccess` here? I think in `hiddenList` it is duplicated join keys. What does it mean to mark them as qualified access.

**Review by viirya — COMMENTED** [review]
All test passed. So I think the new `markAsQualifiedAccess` works in practice, at least for nature join case. Although I have a question about how to interpret it in its semantics.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/MetadataColumnSuite.scala [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+      )
+
+      checkAnswer(
+        dfQuery,
+        Seq(Row(1, "a", 0, "3/1"), Row(2, "b", 0, "0/2"), Row(3, "c", 0, "1/3"))
+      )
+    }
+  }
+
+  test("SPARK-34923: propagate metadata columns through Project") {
```
Comment (huaxingao) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/MetadataColumnSuite.scala:128]: nit: should this be `test("SPARK-40149: propagate metadata columns through Project for  DataFrame API")`?

Shall we keep the 
```
      assertThrows[AnalysisException] {
        sql(s"SELECT index, _partition from (SELECT id, data FROM $t1)")
      }
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3528]: how about `markAsQualifiedAccessOnly`?

File: sql/core/src/test/scala/org/apache/spark/sql/connector/MetadataColumnSuite.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/MetadataColumnSuite.scala:128]: It is kept, in https://github.com/apache/spark/pull/37758/files#diff-0d2dd492482083c6ff24c2572eacf9eeef3f01a180f97fec4dc4c5e08e9fbfb4R203

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3528]: Hmm, sounds okay

**Review by viirya — APPROVED** [review]

**Review by huaxingao — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks for review, merging to master/3.3!

**Pull request closed without merging**